pkg/ctxlock/watcher.go

```diff
@@ -0,0 +1,35 @@
+package ctxlock
+
+// A Watcher waits on two cancellation sources and makes sure to call the
+// wrapped function as soon as possible.
+type watcher struct {
+	onCancel func()
+	done     chan struct{}
+}
+
+func newWatcher(onCancel func()) *watcher {
+	return &watcher{
+		onCancel: onCancel,
+		done:     make(chan struct{}),
+	}
+}
+
+// Watch on the provided channel.
+func (w *watcher) Watch(ch <-chan struct{}) {
+	if ch == nil {
+		panic("nil channel")
```

Comment:
The panic here seems a little strong for an importable package. I get that it's going to be always called in a goroutine so returned errors aren't so useful, but if the channel is `nil` I have no problem with that just `return`ing.

Reply:
My argument would be that this is a private function and it's a programmer error if this is called with a nil channel.

Reply:
Sure, but it's fairly standard practice to de-reference a channel (as happens here in `reconnect()`) I'm not saying it's not safe, I think the cond lock probably covers it, just making the point that unlike writing to a closed channel or reading from a non-existent index, this is a situation where there are other options.

I'm happy for it to stay as is, the struct being private ties the logic up in the package.